### DISCUSSION TITLE
Option to split JSON payload into individual topics?

### DISCUSSION THREAD

**Discussion by nikolac** [arendst/Tasmota#16432] [Q&A] [answered]
Having individual topics for payload attributes would greatly simplify my hub configurations by eliminating duplicate parsing logic. It also allows for more granular subscriptions. 

**Example Neopixel**
Default behavior:
`stat/NEOPIXEL01/RESULT = {"POWER":"OFF","Dimmer":0,"Color":"0,0,0","HSBColor":"46,100,0","Channel":[0,0,0]}`

Execute hypothetical new command
`SplitPayloadIntoTopics 1`

New behavior:
```
stat/NEOPIXEL01/RESULT/POWER = OFF
stat/NEOPIXEL01/RESULT/Dimmer = 0
stat/NEOPIXEL01/RESULT/Color = 0,0,0
stat/NEOPIXEL01/RESULT/HSBColor = 46,100,0
stat/NEOPIXEL01/RESULT/Channel = 0,0,0   
```

**Example with nested objects**
```
# Before
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived = {"PIR03":{"Device":"0xE390","Name":"PIR03","0500<00":"000000140000","ZoneStatusChange":0,"ZoneStatusChangeZone":20,"Occupancy":0,"Endpoint":1,"LinkQuality":74}}

# After
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/Device = 0xE390
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/Name = PIR03
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/0500!00 = 000000140000
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/ZoneStatusChange = 0
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/ZoneStatusChangeZone = 20
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/Occupancy = 0
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/Endpoint = 1
tele/ZIGBEE_BRIDGE/PIR03/ZbReceived/PIR03/LinkQuality = 74
```

A similar inquiry was kinda shot down before https://github.com/arendst/Tasmota/discussions/11069, so I'm not asking for this to be prioritized. 

If I wanted to build this myself, where in the code should I start looking? Hoping there is some kind of publish interceptor.

Also, this would be super helpful with the zigbee bridge, and https://tasmota.github.io/docs/Zigbee/#advanced-topic-sending-sensor-values-to-separated-mqtt-topics kind of eludes to the fact that this is already being thought about.

Limitations in rule count, rule size, make this approach unusable for anything more than one-off scenarios. However, more wildcarding features in the rule definitions might be a good workaround
```
# Proposed wildcarding to make rule more generic
Rule1
   on zbreceived#%device%#%valueName% do publish stat/%device%/%valueName% %value% endon
```

**Comment by s-hadinger** [comment, marked as answer]
Sorry but this won't happen in the core Tasmota because it would generate too much traffic on Mqtt. It is generally preferable to keep a message count low and encode in json

Why can't you do this at Mqtt gateway level?

**Reply by nikolac** [reply]
> Why can't you do this at Mqtt gateway level?

By that, do you mean writing some kind of an external transformation service that subs/parses/publishes to different topics? Yeah, that is a possibility, but I would prefer to address it at the source.

> it would generate too much traffic on Mqtt

Off by default. And/or make it configurable per topic `SplitPayloadIntoTopics tele/ZIGBEE_BRIDGE 1`

> won't happen in the core Tasmota

Understanding the potential risks to overloading mqtt broker, and the very low likelyhood of this ever getting merged, I'd still like to try making it happen as a personal project. 

IF you were to try to do this, what would fit best architecturally? You think introducing a way to inject kind of interceptor https://github.com/arendst/Tasmota/blob/3220fdae449ffdfc1f816f0bbb42adfd796c27be/tasmota/tasmota_xdrv_driver/xdrv_02_9_mqtt.ino#L639 here would make sense? It should take the same arguments by refrence (allowing for transformations), and output true/false flag to skip the rest of the publish. 

For my use case, I would inject an interceptor that splits the payload, recursively publishes, then returns false to prevent the primary call from publishing the original JSON object. 

Thoughts?

**Reply by sfromis** [reply]
While it is not what you ask for, you do have `SetOption4` to get more detailed topics than `.../RESULT` for commands.

More generally, I'd recommend you to get a good JSON parser, which makes it *very easy* to deal with the JSON-structured payloads. IMO, this is much preferable to getting each field separated into their own messages. Such a format directly makes it *harder* to deal with combinations of data ....

If you really want an interceptor to atomize payloads, I'd recommend to not do it in each Tasmota device, but create one MQTT client subscribing to the payloads, and republishing in a different format. With ESP32 and the Berry language, you could even have such an interceptor running in Tasmota and transform data, as Berry works excellently with JSON data.

**Reply by s-hadinger** [reply]
My fear is that some json payloads contain tend of fields, most of which might 'or interest you. Enabling this feature will flood the Mqtt gateway. 

Good point also that it makes it very difficult to detect combination of data, which happens for example in Zigbee

**Reply by nikolac** [reply]
I definitely understand the concern for flooding, and the potential issue of splitting truly coupled updates. Both very strong arguments for not touching _**default**_ behavior. However, I have several use cases where attribute quantity and update frequency are low (minor performance concerns), and I KNOW I don't have an issue with coupled updates, so having the _option_ for more atomic subscriptions would significantly improve my experience. 

I really appreciate both of you educating me on the potential downsides, and why this doesn't make sense for the platform as a whole, but I'm hoping we can speak in the context of a hypothetical world where MQTT performance isn't a concern, and there is no need to correlate atomic updates.

**Reply by sfromis** [reply]
With open source, you are on your own to make the software worse, if you feel a need for that. Play around and see how it works out.

**Reply by nikolac** [reply]
> A similar inquiry was kinda shot down before https://github.com/arendst/Tasmota/discussions/11069, so I'm not asking for this to be prioritized.
>
> If I wanted to build this myself, where in the code should I start looking? Hoping there is some kind of publish interceptor.

I completely get that I'm on my own to build and support this kind of feature, I'm just hoping to leverage the community's experience on the "right" way to get started. 

To follow existing patterns, seems like I should look at the implementation of SetOption4? Appreciate the hint.

And, look at it this way, it can't hurt having another developer in the community semi-comfortable touching the codebase. My motivation to jump in may be selfish, but it opens the door to helping with mainline contributions.

**Reply by s-hadinger** [reply]
I'm still not convinced that we should add to the codebase a feature that drives what I consider an anti-pattern.

Anyways, you can try to add it, but it will requires some significant work to parse the content published to MQTT in a recursive manner, re-publish all the sub-content and make sure that you don't end up in an infinite loop (i.e. re-parsing what you sent yourself).

I still don't understand why you don't this externally, like an ESP32 for example. This way you can subscribe to any topic and re-publish whatever makes sense, and ignore what you don't need. This seems way more flexible and safe. My $.02

**Reply by nikolac** [reply]
Thank you for your feedback, and your guidance. It's nice to see a project with such passionate, protective, and responsive* maintainers.
